ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/UpdateToDbDtoSpec.scala

```diff
@@ -0,0 +1,1191 @@
+          submitters = Some(submitterInfo.actAs.toSet),
+          contract_id = exerciseNode.targetCoid.coid,
+          template_id =
+            Some(createNode.templateId.toString), // taken from explicit divulgedContracts
+          tree_event_witnesses = Set("divulgee"), // taken from explicit blinding info
+          create_argument = Some(emptyArray), // taken from explicit divulgedContracts
+          create_argument_compression = compressionAlgorithmId,
+          event_sequential_id = 0,
+        ),
+        DbDto.CommandCompletion(
+          completion_offset = someOffset.toHexString,
+          record_time = update.recordTime.toInstant,
+          application_id = submitterInfo.applicationId,
+          submitters = submitterInfo.actAs.toSet,
+          command_id = submitterInfo.commandId,
+          transaction_id = Some(update.transactionId),
+          status_code = None,
+          status_message = None,
+        ),
+      )
+    }
+    "handle TransactionAccepted (rollback node)" in {
+      // Transaction
+      // └─ #1 Rollback
+      //    ├─ #2 Create
+      //    └─ #3 Exercise (divulges #2 to divulgee)
+      // - Create and Exercise events must not be visible
+      // - Divulgence events from rolled back Exercise/Fetch nodes must be visible
+      val submitterInfo = someSubmitterInfo
+      val transactionMeta = someTransactionMeta
+      val builder = new TransactionBuilder()
+      val rollbackNode = builder.rollback()
+      val createNode = builder.create(
+        id = builder.newCid,
+        template = "pkgid:M:T",
+        argument = Value.ValueUnit,
+        signatories = List("signatory"),
+        observers = List("observer"),
+        key = None,
+      )
+      val exerciseNode = builder.exercise(
+        contract = createNode,
+        choice = "someChoice",
+        consuming = true,
+        actingParties = Set("signatory"),
+        argument = Value.ValueUnit,
+        result = Some(Value.ValueUnit),
+        choiceObservers = Set("divulgee"),
+        byKey = false,
+      )
+      val rollbackNodeId = builder.add(rollbackNode)
+      builder.add(createNode, rollbackNodeId)
+      builder.add(exerciseNode, rollbackNodeId)
+      val transaction = builder.buildCommitted()
+      val update = Update.TransactionAccepted(
+        optSubmitterInfo = Some(submitterInfo),
+        transactionMeta = transactionMeta,
+        transaction = transaction,
+        transactionId = TransactionId.assertFromString("TransactionId"),
+        recordTime = someRecordTime,
+        divulgedContracts = List.empty,
+        blindingInfo = None,
+      )
+      val dtos = UpdateToDbDto(someParticipantId, valueSerialization, compressionStrategy)(
+        someOffset
+      )(update).toList
+
+      dtos should contain theSameElementsInOrderAs List(
+        // TODO append-only: Why is there a divulgence event? The divulged contract doesn't exist because it was rolled back.
```

Comment:
Should we open a ticket to track this?

Reply:
This particular issue doesn't come from the transaction traversal logic. It's the `BlindingInfo` class that produces a divulgence entry. There is even a [unit test for BlindingInfo](https://github.com/digital-asset/daml/blob/main/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/BlindingSpec.scala#L195) that handles divulgence under rollback nodes.

@moritzkiefer-da @remyhaemmerle-da Is it intended that `BlindingInfo` produces divulgence entries that reference create events rolled back in the same transaction? Is there any point in storing such divulgence events (maybe for a future where divulgence events are directly observable through the ledger API)?

Reply:
`BlindingInfo` included transient contracts already so contracts that are created and archived in the same transaction. Given that, it seemed natural to also include rolled-back creates. If you want to filter out transient and rolled-back creates you can use `inactiveContracts` https://github.com/digital-asset/daml/blob/729afa8c7b68580b6c4aa784e0f2428960eb0635/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala#L406

Reply:
A divulged transient contract should be visible over the transaction tree stream, so we can't filter out divulgence based on `inactiveContracts`. Divulging a rolled back contract on the other hand is something that "never happened" in my mental model.

If the semantics are not clear, I would opt for storing the divulgence event. Divulging rolled back contracts should be something rare, and we can always delete these events later if we don't need them.

Reply:
> A divulged transient contract
I think you’re mixing up divulgence and disclosure here. Disclosure is what lets us see nodes on the ledger API, e.g., the create and exercise for a transient contract. At the moment we do not need this for rollback nodes but we will need it once they are exposed and ignoring the ledger API, you need it in the sense that if you look at privacy between participants those nodes will be shipped around (and this matters in Canton).

Divulgence (from an indexer pov) is what lets us fetch a contract in a later transaction (the fetch node is never exposed on the ledger API). You don’t need that for transient contracts and you don’t need it for rolled back contracts.

Reply:
You can still have a single transaction where a contract is created, then divulged through a fetch/exercise node, then archived, no? _Right now_ such a divulgence event has no use in the index database (and maybe it never will), but I would not change whether such events are stored as part of this PR. I have opened a separate issue to discuss this.

Reply:
As far as I know https://github.com/digital-asset/daml/blob/8747b3d7dcb172705353b59ad54648ca3875c1c0/ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionIndexing.scala#L221 filters out such divulence events for both transient & rolled back nodes but my understanding of the indexer is very superficial.

Reply:
That's the old indexer, the new one takes divulgences directly from the BlindingInfo: https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/scala/platform/store/backend/UpdateToDbDto.scala#L241